### PR TITLE
frozendict --> immutabledict

### PR DESCRIPTION
**Pull request by jkwashbourne**
frozendict deprecated, project abandoned
immutabledict forked from frozendict to replace

### CODE REVIEW

File: .github/workflows/pytest-gpu.yml [hunk 1]
```diff
@@ -81,6 +81,8 @@ jobs:
               git fetch
               git checkout ${{ steps.extract_branch.outputs.branch }}
           fi
+          # Ensure any new requirements are installed
```
Comment (georgebisbas) [inline .github/workflows/pytest-gpu.yml:84]: This is a bit of a hack as we actually need to tweak the image used for testing. It is good for us to think how we treat these cases as a new image will be needed every time we update reqs.

Comment (rhodrin) [reply .github/workflows/pytest-gpu.yml:84]: It's not hack, it's just ensuring the correct requirements are installed if they get changed. Saying it's a hack is a bit like saying 'git pull' is a hack.

Comment (rhodrin) [reply .github/workflows/pytest-gpu.yml:84]: At some point yes, but only when an update occurs that requires to modify something fundamental on the image.

**Comment by rhodrin** [comment]
@jkwashbourne This will now need a rebase owing to the changes to CI-gpu.

**Comment by FabioLuporini** [comment]
closing as superseded by #1278 

we agreed (offline) that the way to go is to rather take control of `frozendict` , which is a tiny data structure, rather than keeping it as a dependence

**Pull request closed without merging**